sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBFileManager.scala

```diff
@@ -0,0 +1,165 @@
+  }
+
+  def readFromFile(metadataFile: File): RocksDBCheckpointMetadata = {
+    val reader = Files.newBufferedReader(metadataFile.toPath, UTF_8)
+    try {
+      val versionLine = reader.readLine()
+      if (versionLine != s"v$VERSION") {
+        throw new IllegalStateException(
+          s"Cannot read RocksDB checkpoint metadata of version $versionLine")
+      }
+      Serialization.read[RocksDBCheckpointMetadata](reader)
+    } finally {
+      reader.close()
+    }
+  }
+
+  def apply(rocksDBFiles: Seq[RocksDBImmutableFile], numKeys: Long): RocksDBCheckpointMetadata = {
+    val sstFiles = rocksDBFiles.collect { case file: RocksDBSstFile => file }
+    val logFiles = rocksDBFiles.collect { case file: RocksDBLogFile => file }
+
+    RocksDBCheckpointMetadata(sstFiles, logFiles, numKeys)
+  }
+}
+
+/**
+ * A RocksDBImmutableFile maintains a mapping between a local RocksDB file name and the name of
+ * its copy on DFS. Since these files are immutable, their DFS copies can be reused.
+ */
+sealed trait RocksDBImmutableFile {
+  def localFileName: String
+  def dfsFileName: String
+  def sizeBytes: Long
+
+  /**
+   * Whether another local file is same as the file described by this class.
+   * A file is same only when the name and the size are same.
+   */
+  def isSameFile(otherFile: File): Boolean = {
+    otherFile.getName == localFileName && otherFile.length() == sizeBytes
+  }
+}
+
+/**
+ * Class to represent a RocksDB SST file. Since this is converted to JSON,
+ * any changes to these MUST be backward-compatible.
+ */
+private[sql] case class RocksDBSstFile(
+    localFileName: String,
+    dfsSstFileName: String,
+    sizeBytes: Long) extends RocksDBImmutableFile {
+
+  override def dfsFileName: String = dfsSstFileName
+}
+
+/**
+ * Class to represent a RocksDB Log file. Since this is converted to JSON,
+ * any changes to these MUST be backward-compatible.
+ */
+private[sql] case class RocksDBLogFile(
+    localFileName: String,
+    dfsLogFileName: String,
+    sizeBytes: Long) extends RocksDBImmutableFile {
+
+  override def dfsFileName: String = dfsLogFileName
+}
+
+object RocksDBImmutableFile {
+  val SST_FILES_DFS_SUBDIR = "SSTs"
+  val LOG_FILES_DFS_SUBDIR = "logs"
+  val LOG_FILES_LOCAL_SUBDIR = "archive"
+
+  def apply(localFileName: String, dfsFileName: String, sizeBytes: Long): RocksDBImmutableFile = {
+    if (isSstFile(localFileName)) {
+      RocksDBSstFile(localFileName, dfsFileName, sizeBytes)
+    } else if (isLogFile(localFileName)) {
+      RocksDBLogFile(localFileName, dfsFileName, sizeBytes)
+    } else {
+      null
+    }
+  }
+
+  def isSstFile(fileName: String): Boolean = fileName.endsWith(".sst")
+
+  def isLogFile(fileName: String): Boolean = fileName.endsWith(".log")
+
+  private def isArchivedLogFile(file: File): Boolean =
+    isLogFile(file.getName) && file.getParentFile.getName == LOG_FILES_LOCAL_SUBDIR
+
+  def isImmutableFile(file: File): Boolean = isSstFile(file.getName) || isArchivedLogFile(file)
```

Comment:
Hmm, so if a log file not archived, it is not immutable?

Reply:
That's right. This is because we only generate the RocksDBLogFile object in the operation of checkpointing. Let's keep this comment open and I'll link this with the caller side for full context. (The caller side is the next PR as my plan - the save path of RocksDBFileManager).

Reply:
I read the save path, but still a bit confused. Is there a log file which is not archived log file? I mean why only archived log file is immutable? `isArchivedLogFile` considers a log file is archived by looking at its parent dir, so when we move a log file to `LOG_FILES_LOCAL_SUBDIR` and make it immutable?

Reply:
```
Is there a log file which is not archived log file?...so when we move a log file to LOG_FILES_LOCAL_SUBDIR and make it immutable?
```
Yes, per the [example](https://github.com/facebook/rocksdb/wiki/Write-Ahead-Log#life-cycle-of-a-wal), the archival is triggered when the flushing happens. The save path is used for syncing the necessary files to DFS, and the save path is called in the commit operation of a state store provider.

Reply:
Yea, this is somehow confusing part to me. Do we have a change to touch log files before they are archived?

Reply:
No, we won't change any files generated by Rocksdb, just sync them with DFS.

Reply:
So there won't be any non-archived log files touched by this. Why we need to distinguish archived log files?

Reply:
Do you mean we sync up non-archived log files in `localOtherFiles` to DFS?

```scala
val (localImmutableFiles, localOtherFiles) = allCheckpointFiles.partition(isImmutableFile)
...
zipToDfsFile(localOtherFiles :+ metadataFile, dfsBatchZipFile(version))
```

Reply:
```
Why we need to distinguish archived log files?
```

Because we'll build a mapping between archived log files and versions. https://github.com/apache/spark/pull/32582/files#diff-e3d3914d0398d61fdd299b1f8d3e869ec6a86e97606677c724969e421c9bf44eR135

```
Do you mean we sync up non-archived log files in localOtherFiles to DFS?
```
Yes. The DFS directory structure was explained in the doc and code comment [here](https://github.com/apache/spark/pull/32582/files#diff-e3d3914d0398d61fdd299b1f8d3e869ec6a86e97606677c724969e421c9bf44eR72)

Reply:
The DFS directory structure only explains SST files, but doesn't mention log files. Does `metadata` also  contain mapping between `archive/00008.log` and `logs/00008-[uuid3].log`? Are archived log files also shared across checkpoint versions?

Reply:
Looks they are shared too as they are immutable files. Can you also update the code comment there?

Reply:
```
Looks they are shared too as they are immutable files. Can you also update the code comment there?
```
Yes that's right. All the immutable files (sst and log files) are contained in checkpoint metadata and shared across checkpoint versions. The un-archived log files are only contained in the ${version}.zip and mapped with a specific checkpoint version.

I think the structure already contains log files. Do you mean I need to emphasize it in the comments (here)[https://github.com/apache/spark/pull/32582/files#diff-e3d3914d0398d61fdd299b1f8d3e869ec6a86e97606677c724969e421c9bf44eR74-R77]?

Reply:
I saw archived log files in the structure, but the comment only mentions SST files. So I thought if it is different than SST files. I think it is better to  mention archived log files along with SST files in the comments.

Reply:
Ah yeah. I know why you feel confused. Make sense, I'll emphasize the "sst files and log files" together in the comments to avoid making others confuse. Thanks!

Reply:
The comment changed in https://github.com/apache/spark/pull/32582/files#diff-e3d3914d0398d61fdd299b1f8d3e869ec6a86e97606677c724969e421c9bf44eR77-R80